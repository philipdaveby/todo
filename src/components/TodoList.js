import React, { useState, useEffect } from 'react';
import TodoCard from './TodoCard';
import InputField from './InputField';

const TodoList = () => {
  const [todos, setTodos] = useState('');
  const [counter, setCounter] = useState(todos.length);

  useEffect(() => () => {
    setTodos('');
  }, []);

  const addTodo = todo => {
    if (!todo.title) {
      const errorTag = document.querySelector('.main__form__error-message');
      errorTag.textContent = 'It is required to set a Title';
      setTimeout(() => {
        errorTag.textContent = '';
      }, 5000);
      return;
    }
    setTodos(prevState => [...prevState, todo]);
    setCounter(prevState => prevState + 1);
  };

  const updateDone = id => {
    const todoIndex = todos.findIndex(todo => todo.key === id);
    todos[todoIndex].done = !todos[todoIndex].done;
    document.querySelector(`li[id="${id}"`).classList.toggle('main__todo__card--done');

    setTodos(todos);
  };

  const removeCard = id => {
    const updatedArr = todos.filter(todo => todo.key !== id);

    setTodos(updatedArr);
  };

  return (
    <main className="main__todolist">
      <section className="main__form">
        <h2 className="main__form__h2">Register new ToDo</h2>
        <InputField addTodo={addTodo} todoKey={counter} />
        <p className='main__form__error-message'></p>
      </section>
      <section className="main__todo">
        <ul className="main__todo__cards">
        { todos ? todos.map(todo => (
        <li key={todo.key} id={todo.key} className={'main__todo__card' } onClick={() => updateDone(todo.key)}>
          <TodoCard cardKey={todo.key} removeCard={removeCard} cardTitle={todo.title}
          cardContent={todo.content} />
        </li>)) : '' }
        </ul>
      </section>
    </main>
  );
};

export default TodoList;
