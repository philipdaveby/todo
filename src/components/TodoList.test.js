import TodoList from './TodoList';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

describe('The TodoList component', () => {

  test('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<TodoList />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has an h2 tag with text "Register new ToDo"', () => {
    render(<TodoList />);
    expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent('Register new ToDo');
  });

  test('contains a list', () => {
    render(<TodoList />);
    const list = screen.getByRole('list');
    expect(list).toBeInTheDocument();
  });

  test('does not have any list items at start', () => {
    const todoList = shallow(<TodoList />);
    expect(todoList.find('li').exists()).toBeFalsy()
  });

  test('Snapshot', () => {
    const todoListComponent = renderer
      .create(<TodoList />)
      .toJSON();
    expect(todoListComponent).toMatchSnapshot();
  });

});
