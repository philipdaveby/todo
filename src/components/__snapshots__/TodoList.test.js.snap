// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The TodoList component Snapshot 1`] = `
<main
  className="main__todolist"
>
  <section
    className="main__form"
  >
    <h2
      className="main__form__h2"
    >
      Register new ToDo
    </h2>
    <form
      aria-label="form"
      className="main__form__todo"
      onSubmit={[Function]}
    >
      <label
        className="main__form__label"
      >
        Title
        <input
          className="main__form__text-input"
          name="Title"
          onChange={[Function]}
          placeholder="Enter a title.."
          type="text"
          value=""
        />
      </label>
      <label
        className="main__form__label"
      >
        Description
        <input
          className="main__form__text-input"
          name="Description"
          onChange={[Function]}
          placeholder="Enter a description.."
          type="text"
          value=""
        />
      </label>
      <input
        id="input__submit"
        type="submit"
        value="Add"
      />
    </form>
    <p
      className="main__form__error-message"
    />
  </section>
  <section
    className="main__todo"
  >
    <ul
      className="main__todo__cards"
    >
      
    </ul>
  </section>
</main>
`;
